### DISCUSSION TITLE
Question about PacketFilterBase canPushPacket implementation

### DISCUSSION THREAD

**Discussion by haug-den-lucas** [inet-framework/inet#964] [General]
Hello,

I'm currently working with some Per-Stream Filtering and Policing in INET and stumbled upon the following lines of code:

https://github.com/inet-framework/inet/blob/61410b46d87c5c4f337d70fa1b3c268425c7dd03/src/inet/queueing/base/PacketFilterBase.cc#L65-L71

In the no-backpressure case, this code accepts all packets in case its consumer accepts all packets.
However, if the consumer does not accept packets, canPushPacket returns true iff the packet criteria does not match, meaning it accepts all packets it will immediately filter out.
I cannot think about any scenario why this implementation would be useful.

Could you please clarify why it is implemented like this?

**Comment by levy** [comment]
The conditions are derived from the following rules:
 - a packet cannot be sent out from the filter module if it doesn't matches the filter
 - if a packet is pushed into the filter using pushPacket(), then only one of two things can happen, it is either deleted or sent out immediately

So if backpressure is not allowed, then a packet that doesn't match the filter can be pushed into the module, but it will be deleted.
For example, you could have a source module that produces all kinds of packets, some of which are not desired. You can simply put a filter module after the source that filters out unwanted packets, effectively creating a compound source module that only generates desired packets.

Does this answer your question?

**Reply by haug-den-lucas** [reply]
Yes, I think that makes sense.
Thanks for clarifying.
